src/FsAutoComplete.Core/TestAdapter.fs

```diff
@@ -0,0 +1,446 @@
+
+    and visitBinding parent b =
+        let (Binding(_, _, _, _, attrs, _, _, pat, _, _, r, _)) = b
+        if isNUnitTest attrs then
+            ident <- ident + 1
+            let entry = {
+                Name = getName pat
+                Range = r
+                Childs = ResizeArray()
+                Id = ident
+                List =  false
+                Type = NUnitType
+            }
+            parent.Childs.Add entry
+
+    let rec visitDeclarations (parent : TestAdapterEntry) decls =
+        for declaration in decls do
+            match declaration with
+            | SynModuleDecl.Let (_, bindings, _) -> for b in bindings do visitBinding parent b
+            | SynModuleDecl.NestedModule (ci, _, decls, _, _) ->
+                let (ComponentInfo(attrs, _, _, ids, _, _, _, r)) = ci
+                let name = String.concat "." [ for i in ids -> i.idText ]
+                ident <- ident + 1
+                let entry = {
+                    Name = name
+                    Range = r
+                    Childs = ResizeArray()
+                    Id = ident
+                    List = true
+                    Type = NUnitType
+                }
+                parent.Childs.Add entry
+                visitDeclarations entry decls
+                if entry.Childs.Count = 0 then
+                    parent.Childs.Remove entry
+                    |> ignore
+            | SynModuleDecl.Types (types, _) -> for t in types do visitTypeDef parent t
+            | _ -> ()
+
+    let visitModulesAndNamespaces parent modulesOrNss =
+        Seq.iter (fun (SynModuleOrNamespace(ids, _, _, decls, _, _, _, r)) ->
+            let name = String.concat "." [ for i in ids -> i.idText ]
+            ident <- ident + 1
+            let entry = {
+                Name = name
+                Range = r
+                Childs = ResizeArray()
+                Id = ident
+                List = true
+                Type = NUnitType
+            }
+            parent.Childs.Add entry
+            visitDeclarations entry decls
+            if entry.Childs.Count = 0 then
+                parent.Childs.Remove entry
+                |> ignore
+        ) modulesOrNss
+
+    let allTests = {
+        Name =  ""
+        Range = Range.range0
+        Childs = ResizeArray()
+        Id = -1
+        List = false
+        Type = ""
+    }
+
+    ast
+    |> Option.iter (function
+        | ParsedInput.ImplFile (ParsedImplFileInput(_, _, _, _, _, modules, _)) -> visitModulesAndNamespaces allTests modules
+        | _ -> ())
+
+    List.ofSeq allTests.Childs
+
+let getXUnitTest ast : TestAdapterEntry list =
+    let mutable ident = 0
+
+    let isXUnitTest (attrs : SynAttributes) =
+        attrs
+        |> List.collect (fun (attr : SynAttributeList) -> attr.Attributes)
+        |> List.exists (fun a ->
+            let str = a.TypeName.Lid |> List.last
+            str.idText.EndsWith "Fact" || str.idText.EndsWith "Factttribute" ||
```

Comment:
typo here: `FactAttribute`

Reply:
I'm personally going to refer to this as `Facttribute` from here on